tests/system_test.py

```diff
@@ -317,19 +322,55 @@ def teardown(self):
+                    self.pid, self.name, msg, ' '.join(self.args),
                     self.outfile + '.cmd', f.read()))
 
+        # If the process is already dead, this might mean that the process has crashed (For example: a router
+        # crash). self.poll() will return a non None value if the process is already gone.
+        # A None value indicates that the process hasn’t terminated yet.
+        # A negative value -N indicates that the child was terminated by signal N (POSIX only).
+        # A positive value indicates a fail.
         status = self.poll()
-        if status is None:      # Still running
+
+        if status is None:
+            # If status is None, it means that the process is still running
+            self.actual = Process.RUNNING
+        elif status > 0:
+            self.actual = Process.EXIT_FAIL
+        else:
+            self.actual = status
+
+        if status is None:
+            # The process is still running.
+            # Call process.terminate() which will try to stop the child by sending it a SIGTERM
+            # but is not guaranteed to stop it
             self.terminate()
-            if self.expect is not None and self.expect != Process.RUNNING:
-                error("still running")
-            self.expect = 0     # Expect clean exit after terminate
-            status = self.wait()
-        if self.expect is not None and self.expect != status:
-            error("exit code %s, expected %s" % (status, self.expect))
+            try:
+                # Wait for TIMEOUT seconds and then catch the
+                # TimeoutExpired and kill() the process.
+                self.wait(timeout=TIMEOUT)
+            except TimeoutExpired:
+                # The terminate call() has failed to terminate the process even after we waited for TIMEOUT seconds.
+                # Send a SIGKILL to the child by calling kill()
+                self._logger.log("Wait timeout expired trying to terminate "
+                                 "process %s (name=%s). Trying to kill process now" % (self.pid, self.name))
+                self.kill()
+                try:
+                    self.wait(timeout=TIMEOUT)
+                    self._logger.log("Process %s killed successfully" % self.pid)
+                except TimeoutExpired:
+                    # There is something seriously wrong if kill() did not kill the process.
+                    # But should the test suite fail because of this ? Right now, it will
+                    # but we need to revisit this.
+                    error_msg = "Error: Could not kill process "
+                    self._logger.log(error_msg + self.pid)
+                    # This will raise a RuntimeError and fail the test
+                    error(error_msg)
+
+        # At this time, we want to check if self.expect and self.actual are the same
+        if self.expect != self.actual:
```

Comment:
What if `self.expect` was None? Then we don't care what self. actual is, IMO.

Reply:
`None` is not allowed for `self.expect`